### PR TITLE
Favorites Feature and Local Notifications

### PR DESCRIPTION
**Pull request by marcosvonstremayr**
In this module I created a Favorites view where you can see all the favorite cards added with the favorite icon in the card detail. When a card is added to favorites a local notification is shown to the user. 

https://user-images.githubusercontent.com/109096201/192598114-daff5eb8-70ad-4d1c-96f6-dadb33335869.mp4

image

### CODE REVIEW

File: lib/src/config/notification_service/local_notification_service.dart [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    return const NotificationDetails(android: androidNotificationDetails);
+  }
+
+  Future<void> showNotification({
+    required int id,
+    required String title,
+    required String body,
+  }) async {
+    final details = await _notificationDetails();
+    await _localNotificationService.show(id, title, body, details);
```
Comment (FranciscoGaraventa) [inline lib/src/config/notification_service/local_notification_service.dart]: trailing comma

File: lib/src/domain/repository/database_response.dart [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        );
+      }
+    } catch (e) {
+      return CardEvent(
+        status: Status.error,
+        errorMsg: "$e",
+      );
+    }
+  }
+}
```
Comment (FranciscoGaraventa) [inline lib/src/domain/repository/database_response.dart]: Add new line at EOF

File: lib/src/domain/repository/favorites_repository.dart [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import '../../data/model/card_model.dart';
+import '../entity/card_event.dart';
+
+abstract class FavoritesRepository {
+  Future<CardEvent> saveFavoriteCard(CardModel card);
+  Future<CardEvent> getFavoriteCard(CardModel card);
+  Future<CardEvent> fetchFavoriteCards();
+}
```
Comment (FranciscoGaraventa) [inline lib/src/domain/repository/favorites_repository.dart]: Add new line at eof

File: lib/src/domain/usecase/get_favorite_usecase.dart [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+class GetFavoriteUsecase extends Usecase<CardEvent, CardModel> {
+  final FavoritesRepository _favoritesRepository;
+
+  GetFavoriteUsecase(this._favoritesRepository);
+
+  @override
+  Future<CardEvent> call({CardModel? params}) async {
+    return await _favoritesRepository.getFavoriteCard(params!);
+  }
+}
```
Comment (FranciscoGaraventa) [inline lib/src/domain/usecase/get_favorite_usecase.dart]: Add new line at eof

File: lib/src/domain/usecase/get_favorites_usecase.dart [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+class GetFavoritesUsecase extends Usecase<CardEvent, CardModel> {
+  final FavoritesRepository _favoritesRepository;
+
+  GetFavoritesUsecase(this._favoritesRepository);
+
+  @override
+  Future<CardEvent> call({CardModel? params}) async {
+    return await _favoritesRepository.fetchFavoriteCards();
+  }
+}
```
Comment (FranciscoGaraventa) [inline lib/src/domain/usecase/get_favorites_usecase.dart]: Add new line at eof

File: lib/src/presentation/bloc/favorites_bloc/favorite_bloc.dart [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+abstract class FavoritesBloc extends Bloc{
+  Stream<CardEvent> getStream();
+  Future<void> saveFavoriteCard(CardModel card);
+  Future<void> isCardFavorite(CardModel card);
+  @override
+  void initialize();
+  @override
+  void dispose();
+}
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/bloc/favorites_bloc/favorite_bloc.dart]: Add new line at eof

File: lib/src/presentation/bloc/favorites_bloc/favorites_list_bloc.dart [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+import '../../../domain/entity/card_event.dart';
+
+abstract class FavoritesListBloc extends Bloc{
+  Stream<CardEvent> getStream();
+  Future<void> getFavoriteCards();
+  @override
+  void initialize();
+  @override
+  void dispose();
+}
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/bloc/favorites_bloc/favorites_list_bloc.dart]: Add new line at eof and format file

File: lib/src/presentation/view/card_detail.dart [hunk 8]
```diff
@@ -136,37 +155,76 @@ class _CardDetailState extends State<CardDetail> {
...
+                                  : Icons.favorite_border,
+                              color: Constants.likeColor,
+                              size: DimensionsConstants.likeIconSize,
+                            ),
+                            onTap: () async {
+                              await widget.favoriteBloc.saveFavoriteCard(cardInfo);
+                              isFavorite
+                                  ? await widget.service.showNotification(
+                                      id: counter,
+                                      title: cardInfo.name! +
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/view/card_detail.dart]: Use string interpolation

File: lib/src/presentation/view/card_detail.dart [hunk 9]
```diff
@@ -136,37 +155,76 @@ class _CardDetailState extends State<CardDetail> {
...
+                                          StringConstants
+                                              .favoritesRemovedNotificationTitle,
+                                      body: StringConstants
+                                          .favoritesRemovedNotificationBody,
+                                    )
+                                  : await widget.service.showNotification(
+                                      id: counter,
+                                      title: cardInfo.name! +
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/view/card_detail.dart]: String interpolation

File: lib/src/config/notification_service/local_notification_service.dart (hunk 1 above)
Comment (marcosvonstremayr) [reply lib/src/config/notification_service/local_notification_service.dart]: done

File: lib/src/domain/repository/database_response.dart (hunk 2 above)
Comment (marcosvonstremayr) [reply lib/src/domain/repository/database_response.dart]: done

File: lib/src/domain/repository/favorites_repository.dart (hunk 3 above)
Comment (marcosvonstremayr) [reply lib/src/domain/repository/favorites_repository.dart]: done

File: lib/src/domain/usecase/get_favorite_usecase.dart (hunk 4 above)
Comment (marcosvonstremayr) [reply lib/src/domain/usecase/get_favorite_usecase.dart]: done

File: lib/src/domain/usecase/get_favorites_usecase.dart (hunk 5 above)
Comment (marcosvonstremayr) [reply lib/src/domain/usecase/get_favorites_usecase.dart]: done

File: lib/src/presentation/bloc/favorites_bloc/favorite_bloc.dart (hunk 6 above)
Comment (marcosvonstremayr) [reply lib/src/presentation/bloc/favorites_bloc/favorite_bloc.dart]: done

File: lib/src/presentation/view/card_detail.dart (hunk 8 above)
Comment (marcosvonstremayr) [reply lib/src/presentation/view/card_detail.dart]: done

File: lib/src/presentation/view/card_detail.dart (hunk 9 above)
Comment (marcosvonstremayr) [reply lib/src/presentation/view/card_detail.dart]: done

**Review by FranciscoGaraventa — APPROVED** [review]

**Review by MarcosCGlobant — APPROVED** [review]

**Review by Juliana-Dimeglio — APPROVED** [review]

**Pull request merged**